make-subcanon-maps.c

```diff
@@ -0,0 +1,218 @@
+#include <stdio.h>
+#include <sys/file.h>
+#include <sys/mman.h>
+#include <unistd.h>
+#include "misc.h"
+#include "tinygraph.h"
+#include "blant.h"
+
+
+//needed for blant stuff
+#define maxBk (1 << (maxK*(maxK-1)/2)) // maximum number of entries in the canon_map
+#define CANON_DIR "canon_maps"
+
+// The following is the most compact way to store the permutation between a non-canonical and its canonical representative,
+// when k=8: there are 8 entries, and each entry is a integer from 0 to 7, which requires 3 bits. 8*3=24 bits total.
+// For simplicity we use the same 3 bits per entry, and assume 8 entries, even for k<8.  It wastes memory for k<4, but
+// makes the coding much simpler.
+typedef unsigned char kperm[3]; // The 24 bits are stored in 3 unsigned chars.
+
+static unsigned int Bk, _k; // _k is the global variable storing k; Bk=actual number of entries in the canon_map for given k.
+
+// Here's where we're lazy on saving memory, and we could do better.  We're going to allocate a static array
+// that is big enough for the 256 million permutations from non-canonicals to canonicals for k=8, even if k<8.
+// So we're allocating 256MBx3=768MB even if we need much less.  I figure anything less than 1GB isn't a big deal
+// these days. It needs to be aligned to a page boundary since we're going to mmap the binary file into this array.
+static kperm Permutations[maxBk] __attribute__ ((aligned (8192)));
+// Here's the actual mapping from non-canonical to canonical, same argument as above wasting memory, and also mmap'd.
+// So here we are allocating 256MB x sizeof(short int) = 512MB.
+// Grand total statically allocated memory is exactly 1.25GB.
+static short int K[maxBk] __attribute__ ((aligned (8192)));
+
+//Needed for BuildGraph
+static TINY_GRAPH *G;
+static int k;
+
+#define MMAP 1 
+// Try to mmap, and if it fails, just slurp in the file (sigh, Windoze)
+void *Mmap(void *p, size_t n, int fd)
+{
+#if MMAP
+    void *newPointer = mmap(p, n, PROT_READ, MAP_PRIVATE|MAP_FIXED, fd, 0);
+    if(newPointer == MAP_FAILED)
+#endif
+    {
+#if !__WIN32__ && !__CYGWIN__ // it will always fail on Windoze so don't bother with a warning
+	Warning("mmap failed");
+#endif
+	if(read(fd, p, n) != n)
+	    Fatal("cannot mmap, or cannot read the file, or both");
+    }
+    return p;
+}
+
+static void BuildGraph(int Gint)
+{
+    int i, j, bitPos=0;
+    int Gint2 = Gint;  // Gint2 has bits nuked as they're used, so when it's zero we can stop.
+    TinyGraphEdgesAllDelete(G);
+#if LOWER_TRIANGLE
+    for(i=k-1;i>0;i--)
+    {
+	for(j=i-1;j>=0;j--)
+#else	// UPPER_TRIANGLE
+    for(i=k-2;i>=0;i--)
+    {
+	for(j=k-1;j>i;j--)
+#endif
+	{
+	    if(!Gint2) break;
+	    int bit = (1 << bitPos);
+	    if(Gint & bit)
+		TinyGraphConnect(G,i,j);
+	    Gint2 &= ~bit;
+	    bitPos++;
+	}
+	if(!Gint2) break;
+    }
+}
+
+#define LOWER_TRIANGLE 1

```

Comment:
Again, this #define should _not_ be duplicated from blant.h; this #define must be a global used consistently by *all* code in the repo, otherwise we risk having inconsistencies.  This cannot be here, it must come from blant.h